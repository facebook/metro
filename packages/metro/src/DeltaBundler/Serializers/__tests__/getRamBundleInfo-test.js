/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @format
 */

'use strict';

const getRamBundleInfo = require('../getRamBundleInfo');

function createModule(name, dependencies, type = 'js/module') {
  return [
    `/root/${name}.js`,
    {
      path: `/root/${name}.js`,
      dependencies: new Map(
        dependencies.map(dep => [
          dep,
          {absolutePath: `/root/${dep}.js`, data: {isAsync: false, name: dep}},
        ]),
      ),
      getSource: () => Buffer.from(`source of ${name}`),
      output: [{type, data: {code: `__d(function() {${name}()});`}}],
    },
  ];
}

const graph = {
  dependencies: new Map([
    createModule('entry', ['foo', 'entry2']),
    createModule('entry2', ['foo2']),
    createModule('foo2', []),
    createModule('foo', ['bar', 'baz', 'qux']),
    createModule('baz', [], 'js/module/asset'),
    createModule('bar', []),
    createModule('qux', []),
  ]),
  importBundleNames: new Set(),
};

const pre = [createModule('pre', [], 'js/script')[1]];

const getRunModuleStatement = moduleId =>
  `require(${JSON.stringify(moduleId)});`;

it('should return the RAM bundle info', async () => {
  expect(
    await getRamBundleInfo('/root/entry.js', pre, graph, {
      processModuleFilter: module => true,
      createModuleId: path => path,
      excludeSource: false,
      getRunModuleStatement,
      getTransformOptions: () => ({
        preloadedModules: {},
        ramGroups: [],
      }),
      dev: true,
      projectRoot: '/root',
      runBeforeMainModule: [],
      runModule: true,
      sourceMapUrl: 'http://localhost/bundle.map',
    }),
  ).toMatchSnapshot();
});

it('should use the preloadedModules and ramGroup configs to build a RAM bundle', async () => {
  const getTransformOptions = async () => ({
    preloadedModules: {'/root/entry2.js': true},
    ramGroups: ['/root/foo.js'],
  });

  const bundleInfo = await getRamBundleInfo('/root/entry.js', pre, graph, {
    processModuleFilter: module => true,
    createModuleId: path => path,
    excludeSource: false,
    getRunModuleStatement,
    getTransformOptions,
    dev: true,
    projectRoot: '/root',
    runBeforeMainModule: [],
    runModule: true,
    sourceMapUrl: 'http://localhost/bundle.map',
  });

  expect(bundleInfo.startupModules.map(({id}) => id)).toEqual([
    '/root/pre.js',
    '/root/entry2.js',
    'require-/root/entry.js',
    'source-map',
  ]);

  expect(bundleInfo.lazyModules.map(({id}) => id)).toEqual([
    '/root/entry.js',
    '/root/foo2.js',
    '/root/foo.js',
    '/root/baz.js',
    '/root/bar.js',
    '/root/qux.js',
  ]);

  expect(bundleInfo.groups).toEqual(
    new Map([
      [
        '/root/foo.js',
        new Set(['/root/bar.js', '/root/baz.js', '/root/qux.js']),
      ],
    ]),
  );
});
