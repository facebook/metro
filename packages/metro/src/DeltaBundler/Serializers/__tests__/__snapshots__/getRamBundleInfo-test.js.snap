// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the RAM bundle info 1`] = `
Object {
  "getDependencies": [Function],
  "groups": Map {},
  "lazyModules": Array [
    Object {
      "code": "__d(function() {entry()},\\"/root/entry.js\\",[\\"/root/foo.js\\",\\"/root/entry2.js\\"],\\"entry.js\\");",
      "id": "/root/entry.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "entry.js",
      "source": "source of entry",
      "sourcePath": "/root/entry.js",
      "type": "js/module",
    },
    Object {
      "code": "__d(function() {entry2()},\\"/root/entry2.js\\",[\\"/root/foo2.js\\"],\\"entry2.js\\");",
      "id": "/root/entry2.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "entry2.js",
      "source": "source of entry2",
      "sourcePath": "/root/entry2.js",
      "type": "js/module",
    },
    Object {
      "code": "__d(function() {foo2()},\\"/root/foo2.js\\",[],\\"foo2.js\\");",
      "id": "/root/foo2.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "foo2.js",
      "source": "source of foo2",
      "sourcePath": "/root/foo2.js",
      "type": "js/module",
    },
    Object {
      "code": "__d(function() {foo()},\\"/root/foo.js\\",[\\"/root/bar.js\\",\\"/root/baz.js\\",\\"/root/qux.js\\"],\\"foo.js\\");",
      "id": "/root/foo.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "foo.js",
      "source": "source of foo",
      "sourcePath": "/root/foo.js",
      "type": "js/module",
    },
    Object {
      "code": "__d(function() {baz()},\\"/root/baz.js\\",[],\\"baz.js\\");",
      "id": "/root/baz.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "baz.js",
      "source": "source of baz",
      "sourcePath": "/root/baz.js",
      "type": "js/module/asset",
    },
    Object {
      "code": "__d(function() {bar()},\\"/root/bar.js\\",[],\\"bar.js\\");",
      "id": "/root/bar.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "bar.js",
      "source": "source of bar",
      "sourcePath": "/root/bar.js",
      "type": "js/module",
    },
    Object {
      "code": "__d(function() {qux()},\\"/root/qux.js\\",[],\\"qux.js\\");",
      "id": "/root/qux.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "qux.js",
      "source": "source of qux",
      "sourcePath": "/root/qux.js",
      "type": "js/module",
    },
  ],
  "startupModules": Array [
    Object {
      "code": "__d(function() {pre()});",
      "id": "/root/pre.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": "pre.js",
      "source": "source of pre",
      "sourcePath": "/root/pre.js",
      "type": "js/script",
    },
    Object {
      "code": "require(\\"/root/entry.js\\");",
      "id": "require-/root/entry.js",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [
          "require-/root/entry.js",
        ],
        "sourcesContent": Array [
          "",
        ],
        "version": 3,
      },
      "name": "entry.js",
      "source": "",
      "sourcePath": "require-/root/entry.js",
      "type": "js/script/virtual",
    },
    Object {
      "code": "//# sourceMappingURL=http://localhost/bundle.map",
      "id": "source-map",
      "map": Object {
        "file": undefined,
        "mappings": "",
        "names": Array [],
        "sources": Array [
          "source-map",
        ],
        "sourcesContent": Array [
          "",
        ],
        "version": 3,
      },
      "name": "source-map",
      "source": "",
      "sourcePath": "source-map",
      "type": "js/script/virtual",
    },
  ],
}
`;
